### DISCUSSION TITLE
[API Request] All entity ids should be guids

### DISCUSSION THREAD

**Discussion by dbrader** [opensubsonic/open-subsonic-api#35] [API requests]
### Type of change

API Clarification

### Proposal description

In the original subsonic server, the entity ids were integers as a string.  Some newer servers have switched to guids.

Can we ask all servers, that haven't already, to switch to using guids?

Guids allow offline-first clients to create and edit entities seamlessly without the need to check-in with the server first before allowing any editing.

### Backward compatibility impact

Backward compatibility should not be broken for those clients that get the id value as a string.  Any client that is not already doing this would be broken on any guid using server and should have patched already.

### Backward compatibility

- [X] No backward compatibility impact.

### API details

Use guids for all entity ids when creating the entity.

### Security impacts

none

**Comment by Tolriq** [comment]
The API says it's a string. What's in the string does not change anything at all to anything.

Can be an int an guid an md5 we should not care as client.
It's the ID it's supposed to be stable for a media.

The rest is server internal and we should not bother them with implementation details.

**Reply by austinried** [reply]
In general I think I agree, but GUID or not, it would be somewhat useful if uniqueness across entities types was guaranteed. Since endpoints like `getCoverArt` exist and use one `id` param for multiple entity types, this should mostly already be the case, but it might be nice if it was made clear across all of them.

**Reply by Tolriq** [reply]
I think there's a misunderstanding and the official doc is not clear, but the id you pass to `getCoverArt` is not the actual id of the item it's the covertArt value you get from the different covertArt returned values. (That should be made more clear in the docs if it's not clear for you I guess)

There's many servers that returns non id values there or the cover art can correspond to something else. You should never ever make any assumption about the expected values of covertArt id. While some servers do support well known ids like pl-xxxx some offer that as a workaround but have other proper ids and other don't. Take  Navidrome the ids are strings and the album covert art are al-albumid_randomvalue.

Like even if you take the original Subsonic API:

http://www.subsonic.org/pages/inc/api/examples/song_example_1.xml

The covertArt value returned for a song is not the id of the song but the id of the album.

There's no reasons to force a server to have unique ids for different kind of items, even if they use guid the risk is near 0% but it's not 0 to have collisions. Many people would use automatic database increments to ensure database stability, forcing a server to handle every id by hand seems bad.
Not talking about the fact that guid storage is not supported by SQLITE for example so it's usage as primary key would be not efficient at all.

Made a PR: https://github.com/opensubsonic/open-subsonic-api/pull/38 the rest of doc already said that the returned covertArt where IDs.

**Reply by austinried** [reply]
> the id you pass to getCoverArt is not the actual id of the item it's the covertArt value you get from the different covertArt returned values. (That should be made more clear in the docs if it's not clear for you I guess)

Oh I know that's the way the original subsonic server is implemented, but the docs have had that wording for years and years, and because of that, many servers also accept the actual ID of an entity, not just the coverArt value.

**Reply by Tolriq** [reply]
Your looking it from the wrong perspective here. The goal of the project is to clear up the docs and improve the base API (So the subsonic one, not what some may have built somewhere). 
If we stick to the docs then we can't download playlist covers then since the ids are pl- and not documented in the doc ? There's many mistakes in the docs but the API is already full of flaws to not add more to it.

The main thing here is that is we enforce the API to be based on broken docs and what some have done we acknowledge your request that each ID should be independent between ALL the entities.
This is a complete non sense from any point of view and you'd break all the servers who implement the API correctly, and use auto increment ids. Forcing them to migrate would mean most would not even bother to join this project. Since this absolutely not something we should force on the servers. Internals implementations are internal this is a open API. Not an open implementation.

So if you want to consider this a breaking change then it's a breaking change either way. 
We can either decide : 
- to validate the bad version that will enforce insane changes to servers (So the actual opposite of this project).
- to clear up the current documentation to have a clean base without being too disruptive. New wording allows current servers to do whatever they want, but allows new servers to actually code this properly. If a client makes wrong assumption on a new server or on all the other servers they do not yet support then users will report to the clients that will their implementation.

Seems we have our first major choice to make
@opensubsonic/servers @opensubsonic/clients your participation is important here.

**Reply by austinried** [reply]
To be honest I think you are misreading me here. I said it **would be nice** to guarantee uniqueness, for all of the well-documented advantages UUIDs have over auto-incrementing integers that I'm not going to get into here, and that I **generally agree** with your original response because I'm not sure it really brings enough advantages to force servers not already using a GUID system to switch to one.

The unfortunate original wording of the `getCoverArt` endpoint was only meant to be an example, I was just describing GUIDs as one of the effects it has likely had on server implementations. I will say though that the way you have framed the two options as "insane" and apparently not insane is really not promoting a good discussion on the topic.

Another option might be to change the wording to something that clarifies that both ID and coverArt are supported, which is weird, but is what the history of the situation has created and would not break existing implementations. Yet another option might be to add another `coverArt` parameter to the endpoint and actually differentiate between ID and coverArt, to "more correctly" support both.  These both imply the server is using a unique ID though, so maybe it would be better to specify `id` is coverArt but also add parameters for `albumId`, `artistId`, `songId`, etc. to match other endpoints, but this would break clients that are currently passing an actual `id`.

But I'm really just throwing out ideas here, I don't really have a horse in this race for what my client uses this endpoint for. I would like to hear if anyone else has strong opinions about how `getCoverArt` should work, especially if any clients are passing `id` or if any severs that are not already think supporting `id` would be a large lift.

**Reply by Tolriq** [reply]
I'm not misreading, I'm explaining the impact of current documentation and the purpose of this project.

**The number 1 purpose of this project is to clarify any possible misunderstanding in the doc that allows servers to behave differently for a given end point.** The second point is to improve the app to add the missing things.

So for something like this there's no **would be nice or should** this a major impacting stuff of the api definition.
Either we enforce the servers to have unique different ids for each media type and then the doc can accept an id that can represent any media type. Or we don't. An ID can not be defined as either unique or not unique this changes everything on server side implementation (and potentially on client side too). New fields or end point can be optional. Not base definition.

- If we don't then we clarify the doc to remove the part that may let people think we do.  That what I think is the correct way to handle this since it's the actual way it was build in original API.
- If we do enforce then we enforce with all the impacts this have on the servers to comply.

Again forcing unique IDs is a major breaking change too with a lot more impacts on current servers. This is mostly against the previous votes that was vastly for nearly non breaking changes instead of building a complete V2. If the ids content / definitions are full part of the API (Not the case currently) then it's the first brick of the API changing so it's better to write a V2 from scratch. Breaking the base but still building on something almost broken would be the worst of all solutions.

Tl;DR; OpenSubsonic API documentation is meant to be without risk of wrong interpretation by clients and servers to avoid the current mess with each servers doing something different. We are in the perfect example here about why this is important.

Edit: By insane I mean starting this project by doing the opposite of what it was decided to be. Simplify servers life by not doing a V2 API. A server announcing being OpenSubsonic compatible will still need to comply with the doc else the project serves no purpose.

**Comment by epoupon** [comment]
> Guids allow offline-first clients to create and edit entities seamlessly without the need to check-in with the server first before allowing any editing.

So if I understand correctly, your use case is to generate stuff client side (for ex. playlists), assign some GUID yourself for each entity and once online create them on the server with your own provided GUIDs?

**Reply by Tolriq** [reply]
That would also implies that clients have control of the IDs, and open consistency / duplicate issues.

**Reply by epoupon** [reply]
Actually, being offline does not prevent this use case, even with the current api.
Client side each entry may have an internal id (whatever you want, guid, incremental counter,...) and a server id (possibly not set until sync). I guess if you can have multiple servers at the same time this is the only option.

**Reply by Tolriq** [reply]
Exactly what I'm doing. There's never the need for clients to impose anything on server IDs.
